include/seahorn/PredicateAbstraction.hh

```diff
@@ -0,0 +1,151 @@
+#include "ufo/Smt/Z3n.hpp"
+#include "ufo/Smt/EZ3.hh"
+#include "seahorn/HornClauseDBWto.hh"
+
+namespace seahorn
+{
+	using namespace llvm;
+
+	class HornDbModel
+	{
+	private:
+		ExprMap relToDefMap;
+		std::map<Expr, ExprVector> currentCandidates;
+	public:
+		HornDbModel() {}
+		void initModelFromFP(HornClauseDB &db, ZFixedPoint<EZ3> &fp);
+		void addDef(Expr fapp, Expr def);
+		Expr getDef(Expr fapp);
+		HornDbModel(ExprMap model) : relToDefMap(model) {}
+		virtual ~HornDbModel(){}
+		void setCurrentCandidates(std::map<Expr, ExprVector> currentCandidates) { this->currentCandidates = currentCandidates; }
+//		ExprMap& getRelToSolutionMap() {return relToDefMap;}
+//		void setRelToSolutionMap(ExprMap map) {relToDefMap = map;}
+	};
+
+	class HornModelConverter
+	{
+	public:
+		// converts a model from one database to another. returns false on failure.
+		virtual bool convert (HornDbModel &in, HornDbModel &out) = 0;
+	};
+
+	class PredAbsHornModelConverter : public HornModelConverter
+	{
+	private:
+		std::map<Expr, ExprMap> relToBoolToTermMap;
+		std::map<Expr, Expr> newToOldPredMap;
+		HornClauseDB* abs_db;
+	public:
+		PredAbsHornModelConverter() {}
+		bool convert (HornDbModel &in, HornDbModel &out);
+		std::map<Expr, ExprMap>& getRelToBoolToTermMap() {return relToBoolToTermMap;}
+		void setNewToOldPredMap(std::map<Expr, Expr> &newToOldMap) {newToOldPredMap = newToOldMap;}
+		void setAbsDB(HornClauseDB &db) {abs_db = &db;}
+	};
+
+	class PredicateAbstractionAnalysis
+	{
+	private:
+	    std::map<Expr, Expr> oldToNewPredMap;
+	    std::map<Expr, Expr> newToOldPredMap;
+	    std::map<Expr, ExprVector> currentCandidates;
+
+	    std::unique_ptr<ufo::ZFixedPoint <ufo::EZ3> >  m_fp;
+	    HornifyModule& m_hm;
+
+	public:
+	    PredicateAbstractionAnalysis(HornifyModule &hm) : m_hm(hm) {}
+	    ~PredicateAbstractionAnalysis() {m_fp.reset (nullptr);}
+
+	    ufo::ZFixedPoint<ufo::EZ3>& getZFixedPoint () {return *m_fp;}
+
+	    void runAnalysis();
+		void guessCandidate(HornClauseDB &db);
+		ExprVector relToCand(Expr fdecl);
+		HornClauseDB generateAbstractDB(HornClauseDB &db, PredAbsHornModelConverter &converter);
+		void generateAbstractRelations(HornClauseDB &db, HornClauseDB &new_DB);
+		void generateAbstractRules(HornClauseDB &db, HornClauseDB &new_DB, PredAbsHornModelConverter &converter);
+		void generateAbstractQueries(HornClauseDB &db, HornClauseDB &new_DB);
+		void printInvars(HornClauseDB &db, HornDbModel &origModel);
+	};
+
+	class PredicateAbstraction : public llvm::ModulePass
+	{
+	public:
+	    static char ID;
+
+	    PredicateAbstraction() : ModulePass(ID) {}
+	    virtual ~PredicateAbstraction() {}
+	    virtual bool runOnModule (Module &M);
+	    virtual void getAnalysisUsage (AnalysisUsage &AU) const;
+	    virtual const char* getPassName () const {return "PredicateAbstraction";}
+	};
+
+	class HornDbUtils
+	{
+	public:
+		static Expr applyArgsToBvars(Expr cand, Expr fapp, std::map<Expr, ExprVector> currentCandidates);
```

Comment:
why static class members? This is C++, we can use functions not wrapped in any class.